src/mapgen.cpp

```diff
@@ -116,7 +116,13 @@ void map::generate( const tripoint &p, const time_point &when )
     //  because other submaps won't be touched.
     for( int gridx = 0; gridx < my_MAPSIZE; gridx++ ) {
         for( int gridy = 0; gridy < my_MAPSIZE; gridy++ ) {
-            setsubmap( get_nonant( { gridx, gridy } ), new submap() );
+            if( zlevels ) {
+                for( int gridz = -OVERMAP_DEPTH; gridz <= OVERMAP_HEIGHT; gridz++ ) {
+                    setsubmap( get_nonant( { gridx, gridy, gridz } ), new submap() );
```

Comment:
The new submaps (those that are not part of the generated map) are not properly delete. See the end of this function:

There new submaps that should be kept (have been generated here) are saved, and the other submaps are delete again. But the loops there only go over X and Y, ignoring the Z-level.

Note that map generates only 2x2 submaps (on a specific z-level), anything outside of that range may already exist.

Why does the mapgen code even access the other z-levels, the maps there may have already been generated!

Reply:
>are not properly delete.

Agreed, it needs to be deleted there `delete get_submap_at_grid( { i, j, p.z } );`. Also probably should set the submap to `nullptr` after deletion.

From the comment in code:
> At the bottom of this function, we save the upper-left 4 submaps, and delete the rest.

Looks like I am missing the point of this generation. Will recheck this function and this fix.

May be it is not required to initialize full gird, but instead add the null-checks for all usages of submaps?

Reply:
I think I have found the origin of this ussue: the submap was inserted always at 0 z-level  `{ gridx, gridy }`, but used, saved and deleted on specific z-level `i, j, p.z`.

Fixed new submap inserting on correct z-level. Added some extra checks.